Classes/common/HTTP/CDTURLSessionTask.m

```diff
@@ -108,16 +104,55 @@ - (NSURLSessionTaskState)state
 
 #pragma mark Helpers
 
+- (void)makeRequest
+{
+    __block CDTHTTPInterceptorContext *ctx =
+        [[CDTHTTPInterceptorContext alloc] initWithRequest:[self.request mutableCopy]];
+
+    // We make sure all objects support `interceptRequestInContext:` during init.
+    for (NSObject<CDTHTTPInterceptor> *obj in self.requestInterceptors) {
+        ctx = [obj interceptRequestInContext:ctx];
+    }
+
+    __weak CDTURLSessionTask *weakSelf = self;
+    self.inProgressTask = [self.session
+        dataTaskWithRequest:ctx.request
+          completionHandler:^void(NSData *data, NSURLResponse *response, NSError *error) {
+            __strong CDTURLSessionTask *strongSelf = weakSelf;
+            if (strongSelf) {
+                ctx.response = response;
+                for (NSObject<CDTHTTPInterceptor> *obj in strongSelf.responseInterceptors) {
+                    ctx = [obj interceptResponseInContext:ctx];
+                }
+
+                if (ctx.shouldRetry && strongSelf.remainingRetries > 0) {
+                    // retry
+                    strongSelf.remainingRetries--;
+                    [strongSelf makeRequest];
+
+                } else if (strongSelf.completionHandler) {
+                    data = [NSData dataWithData:data];
+                    strongSelf.completionHandler(data, response, error);
+                }
+            }
+          }];
+    [self.inProgressTask resume];
+}
+
 /**
- Copy the interceptor array, filtering out non-compliant classes.
+ Spilt interceptor array into two arrays, filtering out non-complient classes as we go
 
  We do this once during -init... to avoid spamming the logs for every retry. This checks for
  responding to `interceptRequestInContext:` as we're creating a request interceptor array,
  not a response one.
+
+ @return an array where the first element are the  request interceptors and the second element is
+ the response interceptors
  */
 - (NSArray *)filterRequestInterceptors:(NSArray *)proposedRequestInterceptors
 {
-    NSMutableArray *acc = [NSMutableArray array];
+    NSMutableArray *responseInterceptors = [NSMutableArray array];
+    NSMutableArray *requestInterceptors = [NSMutableArray array];
```

Comment:
Frankly I'd get rid of the warnings (e.g., you don't get warnings when you add a delegate that doesn't implement methods), and simplify this into two methods one for the request and one for the response. Then you can also avoid the awkward return semantics.